spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java

```diff
@@ -0,0 +1,348 @@
+import org.apache.spark.sql.types.ByteType;
+import org.apache.spark.sql.types.DataType;
+import org.apache.spark.sql.types.DataTypes;
+import org.apache.spark.sql.types.Decimal;
+import org.apache.spark.sql.types.DecimalType;
+import org.apache.spark.sql.types.IntegerType;
+import org.apache.spark.sql.types.LongType;
+import org.apache.spark.sql.types.ShortType;
+import org.apache.spark.sql.types.StringType;
+import org.apache.spark.sql.types.StructField;
+import org.apache.spark.sql.types.StructType;
+import org.apache.spark.unsafe.types.UTF8String;
+
+/**
+ * A Spark function implementation for the Iceberg truncate transform.
+ *
+ * <p>Example usage: {@code SELECT system.truncate(1, 'abc')}, which returns the String 'a'.
+ *
+ * <p>Note that for performance reasons, the given input width is not validated in the
+ * implementations used in code-gen. The width must remain non-negative to give meaningful results.
+ */
+public class TruncateFunction implements UnboundFunction {
+
+  private static void validateTruncationWidthType(DataType widthType) {
+    if (!DataTypes.IntegerType.sameType(widthType)
+        && !DataTypes.ShortType.sameType(widthType)
+        && !DataTypes.ByteType.sameType(widthType)) {
+      throw new UnsupportedOperationException(
+          "Expected truncation width to be one of [ByteType, ShortType, IntegerType], but found "
+              + widthType);
+    }
+  }
+
+  @Override
+  public BoundFunction bind(StructType inputType) {
+    if (inputType.fields().length != 2) {
+      throw new UnsupportedOperationException(
+          "Cannot bind truncate: wrong number of inputs (expected width and value)");
+    }
+
+    StructField widthField = inputType.fields()[0];
+    StructField toTruncateField = inputType.fields()[1];
+    validateTruncationWidthType(widthField.dataType());
+
+    DataType toTruncateDataType = toTruncateField.dataType();
+    if (toTruncateDataType instanceof ByteType) {
+      return new TruncateTinyInt();
+    } else if (toTruncateDataType instanceof ShortType) {
+      return new TruncateSmallInt();
+    } else if (toTruncateDataType instanceof IntegerType) {
+      return new TruncateInt();
+    } else if (toTruncateDataType instanceof LongType) {
+      return new TruncateBigInt();
+    } else if (toTruncateDataType instanceof DecimalType) {
+      return new TruncateDecimal(
+          ((DecimalType) toTruncateDataType).precision(),
+          ((DecimalType) toTruncateDataType).scale());
+    } else if (toTruncateDataType instanceof StringType) {
+      return new TruncateString();
+    } else if (toTruncateDataType instanceof BinaryType) {
+      return new TruncateBinary();
+    } else {
+      throw new UnsupportedOperationException(
+          String.format(
+              "Invalid input type to truncate. Expected one of [ByteType, ShortType, IntegerType, LongType, "
+                  + "DecimalType, StringType, BinaryType], but found %s",
```

Comment:
These names are internal Spark names for types. I think we should use the SQL names instead:

```
Invalid input types: truncate(width=%s, col=%s), expected col to be tinyint, shortint, int, bigint, decimal, string, or binary
```

Reply:
Agree on updating the data types to use the SQL types. But Spark adds some to this error message. Let me show you an example.

When passing `(int, interval year to month)` types, I get the following message now.

```
 "Function 'truncate' cannot process input: (int, interval year to month): Invalid input type to truncate. Expected one of [ByteType, ShortType, IntegerType, LongType, DecimalType, StringType, BinaryType], but found YearMonthIntervalType(0,1)"
```

So I would say that we don't need to repeat the name of the function necessarily as Spark adds it. I'd suggest using 
`expected truncation col to be tinyint, shortint, int, bigint, decimal, string, or binary`, so that the error message will wind up being:

`" Function 'truncate' cannot process input: (int, interval year to month): expected truncation col to be tinyint, shortint, int, bigint, decimal, string, or binary"`

Reply:
TLDR - Spark prepends `Function 'truncate' cannot process input: (int, interval year to month): ` to the error message from the analysis exception.

Reply:
Okay, then we don't need all the context. In that case, just change it to `Expected tinyint, shortint, int, long, decimal, string, or binary: interval year to month `

Reply:
I updated the test error messages to show the full message that spark adds (which is very similar to the one you proposed).